internal/applier/helm_test.go

```diff
@@ -0,0 +1,315 @@
+package applier_test
+
+import (
+	"context"
+	"errors"
+	"os"
+	"testing"
+	"testing/fstest"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	helmclient "github.com/operator-framework/helm-operator-plugins/pkg/client"
+
+	v1 "github.com/operator-framework/operator-controller/api/v1"
+	"github.com/operator-framework/operator-controller/internal/applier"
+)
+
+type mockPreflight struct {
+	installErr error
+	upgradeErr error
+}
+
+func (mp *mockPreflight) Install(context.Context, *release.Release) error {
+	return mp.installErr
+}
+
+func (mp *mockPreflight) Upgrade(context.Context, *release.Release) error {
+	return mp.upgradeErr
+}
+
+type mockActionGetter struct {
+	actionClientForErr error
+	getClientErr       error
+	installErr         error
+	dryRunInstallOK    bool
+	upgradeErr         error
+	dryRunUpgradeOK    bool
+	reconcileErr       error
+	desiredRel         *release.Release
+	currentRel         *release.Release
+}
+
+func (mag *mockActionGetter) ActionClientFor(ctx context.Context, obj client.Object) (helmclient.ActionInterface, error) {
+	return mag, mag.actionClientForErr
+}
+
+func (mag *mockActionGetter) Get(name string, opts ...helmclient.GetOption) (*release.Release, error) {
+	return mag.currentRel, mag.getClientErr
+}
+
+func (mag *mockActionGetter) History(name string, opts ...helmclient.HistoryOption) ([]*release.Release, error) {
+	return nil, mag.getClientErr
+}
+
+func (mag *mockActionGetter) Install(name, namespace string, chrt *chart.Chart, vals map[string]interface{}, opts ...helmclient.InstallOption) (*release.Release, error) {
+	if mag.dryRunInstallOK {
+		mag.dryRunInstallOK = false
+		return mag.desiredRel, nil
+	}
```

Comment:
This only works because the current code always calls:
- Install(dryRun=true)
- Install(dryRun=false)

In that order.

I worry that we could potentially change that guarantee in the future, but in a way where our tests continue to pass when they shouldn't.

It might be worth it to handle the functional `opts` and explicitly check to see if this call is a dry run.

```suggestion
	i := action.Install{}
	for _, opt := range opts {
		if err := opt(&i); err != nil {
			return nil, err
		}
	}
	if i.DryRun && mag.dryRunInstallOK {
		return mag.desiredRel, nil
	}
```

Reply:
Same for upgrade.

Reply:
good idea, the flag with order was a little hacky ;) I'll make the changes for both `Install()` and `Upgrade()`